### PR TITLE
Add the StartEdit.DurationInputted action

### LINKED ISSUE #212: Add the DurationInputted action to the StartEdit reducer
## Description
This action is emitted when the user is inputting valid durations in the manual duration input view from The Wheel™.
- it should change the duration of the time entry being edited `editableTimeEntry` (start time if running, end time otherwise)
- invalid durations should be filtered out

## Definition of done
- [x] The action is added
- [x] If the time entry is running, changing the duration should change the start time of the `editableTimeEntry`
- [x] If the time entry is stopped, changing the duration should change the actual duration of the `editableTimeEntry`
- [x] If the time entry is new, a start time should be set and the time entry should be then handled like a running time entry.
- [x] Invalid durations are filtered out 
- [x] Tests are added
- [x] The [documentation](https://www.notion.so/toggl/DurationInputted-31331ad5d01a45bea600a2075bcd0a16) is updated to reflect the changes done in this PR

### PR DESCRIPTION
**Pull request by daividssilverio**
### Technical considerations
Nothing special going on.

### Why is it done like this and not some other way?
I've changed the reduce function from an expression body in favor of a block body to be able to do early returns. Let me know if you are against it and why.

### Relationships
Closes #212

### Pull Request checklist
- [x] The code does not hinder the app's performance
- [x] The code does not introduce regressions
- [x] The code in this PR is tested thoroughly enough that changing the behavior to an incorrect one breaks tests

### Pointers
Review by commit. You may refer to the iOS implementation as well.

### Review template

- [ ] The code is thoroughly tested in well written and resilient manner
- [ ] The code is performant and leads to an optimal user experience
- [ ] The code is testable and does not cut corners for the sake of delivering
- [ ] The code considers older platforms well and will work in older devices just like it will in new ones

## Merge permissions
🦑 it.

### CODE REVIEW

File: timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.converter.ConvertWith
+import org.junit.jupiter.params.converter.SimpleArgumentConverter
+import org.junit.jupiter.params.provider.ValueSource
+import org.threeten.bp.Duration
+import org.threeten.bp.OffsetDateTime
+import org.threeten.bp.ZoneOffset
+
+@ExperimentalCoroutinesApi
+internal class DurationInputtedActionTests {
```
Comment (heytherewill) [inline timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt:28]: No Display name?

File: timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+
+    init {
+        every { timeService.now() } returns now
+    }
+
+    private fun createEditableTimeEntry(startTime: OffsetDateTime?, duration: Duration?) =
+        EditableTimeEntry.empty(1).copy(startTime = startTime, duration = duration)
+
+    @Test
+    fun `when the editable time entry is new`() = runBlockingTest {
```
Comment (heytherewill) [inline timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt:45]: The test names are not ideal in this class. I'd rather see `does _description of what is done_ when the editable time entry is new`

File: timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+                    startTime = initialStartTime,
+                    duration = initialDuration
+                )
+            )
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(longs = [-1L, 0L, 999L, 1000L])
+    fun `no effects are dispatched`(@ConvertWith(LongToDuration::class) durationInputted: Duration) = runBlockingTest {
```
Comment (heytherewill) [inline timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt:144]: When are no effects dispatched?

Comment (daividssilverio) [reply timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt:144]: The current behaviour is that we simply ignore invalid input in the wheel's numeric input.
So, regularly there's no effect produced, and when invalid input is made, nothing happens.

File: timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt (hunk 1 above)
Comment (daividssilverio) [reply timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt:28]: I can add one, didn't think it was needed.

File: timer/src/test/java/com/toggl/timer/startedit/domain/BillableTappedActionTests.kt [hunk 4]
```diff
@@ -13,7 +14,8 @@ import kotlinx.coroutines.ExperimentalCoroutinesApi
 class BillableTappedActionTests : FreeCoroutineSpec() {
     init {
         val repository = mockk<TimeEntryRepository>()
-        val reducer = StartEditReducer(repository, dispatcherProvider)
+        val timeService = mockk<TimeService>()
+        val reducer = StartEditReducer(repository, timeService, dispatcherProvider)
```
[the remark this answers is not in the document]
Comment (daividssilverio) [reply timer/src/test/java/com/toggl/timer/startedit/domain/BillableTappedActionTests.kt:18]: Sure thing.

File: timer/src/main/java/com/toggl/timer/startedit/domain/StartEditReducer.kt [hunk 5]
```diff
@@ -16,18 +18,20 @@ import com.toggl.timer.common.domain.StartTimeEntryEffect
...
     private val dispatcherProvider: DispatcherProvider
 ) : Reducer<StartEditState, StartEditAction> {
 
     override fun reduce(
         state: MutableValue<StartEditState>,
         action: StartEditAction
-    ): List<Effect<StartEditAction>> =
-        when (action) {
+    ): List<Effect<StartEditAction>> {
+        return when (action) {
```
[the remark this answers is not in the document]
Comment (daividssilverio) [reply timer/src/main/java/com/toggl/timer/startedit/domain/StartEditReducer.kt:34]: to be able to do early returns. Check https://github.com/toggl/android/pull/263/files#diff-e98d0e441959ab420d188af3d5237d1fR112-R113

File: timer/src/test/java/com/toggl/timer/startedit/domain/HelperFunctions.kt [hunk 6]
```diff
@@ -22,4 +26,10 @@ fun createInitialState(
         editableTimeEntry = editableTimeEntry,
         autocompleteSuggestions = autoCompleteSuggestions,
         dateTimePickMode = DateTimePickMode.None
-    )
\ No newline at end of file
+    )
+
+fun mockReducer(
```
Comment (heytherewill) [inline timer/src/test/java/com/toggl/timer/startedit/domain/HelperFunctions.kt:31]: This is a bad name. The reducer is not mocked, only its dependencies are

File: timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt [hunk 7]
```diff
@@ -56,7 +58,7 @@ internal class DurationInputtedActionTests {
     }
 
     @Test
-    fun `when the editable time entry is running`() = runBlockingTest {
+    fun `sets its value to the editableTimeEntry's startTime when the time entry is running`() = runBlockingTest {
```
Comment (heytherewill) [inline timer/src/test/java/com/toggl/timer/startedit/domain/DurationInputtedActionTests.kt:61]: `The DurationInputted action sets its value to the editableTimeEntry's startTime when the time entry is running`?
I think you meant 
```suggestion
    fun `sets the editableTimeEntry's startTime when the time entry is running`() = runBlockingTest {
```

File: timer/src/test/java/com/toggl/timer/startedit/domain/HelperFunctions.kt (hunk 6 above)
Comment (daividssilverio) [reply timer/src/test/java/com/toggl/timer/startedit/domain/HelperFunctions.kt:31]: Alright, I'm gonna be very specific then.

Comment (daividssilverio) [reply timer/src/test/java/com/toggl/timer/startedit/domain/HelperFunctions.kt:31]: Or maybe not, `createReducer` seems enough.

**Pull request merged**